src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java

```diff
@@ -3049,7 +3049,36 @@ ByteVector fromArray(VectorSpecies<Byte> species,
                                    byte[] a, int offset,
                                    int[] indexMap, int mapOffset) {
         ByteSpecies vsp = (ByteSpecies) species;
-        return vsp.vOp(n -> a[offset + indexMap[mapOffset + n]]);
+        IntVector.IntSpecies isp = IntVector.species(vsp.indexShape());
+        Objects.requireNonNull(a);
+        Objects.requireNonNull(indexMap);
+        Class<? extends ByteVector> vectorType = vsp.vectorType();
+
+        int loopIncr = 0;
+        VectorSpecies<Integer> lsp = null;
+
+        // Constant folding should sweep out following conditonal logic.
+        if (isp.length() > IntVector.SPECIES_PREFERRED.length()) {
+            lsp = IntVector.SPECIES_PREFERRED;
+        } else {
+            lsp = isp;
+        }
+
+        // Check indices are within array bounds.
+        for (int i = 0; i < vsp.length(); i += lsp.length()) {
+            IntVector vix = IntVector
+                .fromArray(lsp, indexMap, mapOffset + i)
+                .add(offset);
+            vix = VectorIntrinsics.checkIndex(vix, a.length);
+        }
```

Comment:
This has finished the boundary checks for the index array. But the drawback is it also generates the index vectors (I mean the load instructions), which may be duplicated with intrinsification on some hardwares like SVE (sve gatter instructions need the index vectors as well), because you passed the index address and offset to compiler. So for SVE, we will need double load instructions than what it actually needs.

Reply:
Will GVN not be able to share the matching loads to avoid duplicity.